Flink types serialization codegen error: no such constructor: org.apache.flink.table.data.binary.BinaryStringDataFuryCodec_0.<init>(Fury,Class)void/invokeSpecial

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/fury/issues) and found no similar issues.


### Version

main

### Component(s)

Java

### Minimal reproduce step

None

### What did you expect to see?

java.lang.Error: java.lang.IllegalAccessException: no such constructor: org.apache.flink.table.data.binary.BinaryStringDataFuryCodec_0.<init>(Fury,Class)void/invokeSpecial


### What did you see instead?

```java
java.lang.Error: java.lang.IllegalAccessException: no such constructor: org.apache.flink.table.data.binary.BinaryStringDataFuryCodec_0.<init>(Fury,Class)void/invokeSpecial
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1155) ~[?:1.8.0_372]
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_372]
at java.lang.Thread.run(Thread.java:879) [?:1.8.0_372]
Caused by: java.lang.IllegalAccessException: no such constructor: org.apache.flink.table.data.binary.BinaryStringDataFuryCodec_0.<init>(Fury,Class)void/invokeSpecial
at java.lang.invoke.MemberName.makeAccessException(MemberName.java:867) ~[?:1.8.0_372]
at java.lang.invoke.MemberName$Factory.resolveOrFail(MemberName.java:1003) ~[?:1.8.0_372]
at java.lang.invoke.MethodHandles$Lookup.resolveOrFail(MethodHandles.java:1386) ~[?:1.8.0_372]
at java.lang.invoke.MethodHandles$Lookup.findConstructor(MethodHandles.java:924) ~[?:1.8.0_372]
at org.apache.fury.serializer.Serializers.createSerializer(Serializers.java:127) ~[flash-serializer-v1.jar:?]
at org.apache.fury.serializer.Serializers.newSerializer(Serializers.java:104) ~[flash-serializer-v1.jar:?]
at org.apache.fury.resolver.ClassResolver$1.onSuccess(ClassResolver.java:941) ~[flash-serializer-v1.jar:?]
at org.apache.fury.resolver.ClassResolver$1.onSuccess(ClassResolver.java:938) ~[flash-serializer-v1.jar:?]
at org.apache.fury.builder.JITContext.lambda$registerSerializerJITCallback$0(JITContext.java:94) ~[flash-serializer-v1.jar:?]
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_372]
... 2 more
Caused by: java.lang.LinkageError: bad method type alias: (Fury,Class)void not visible from class org.apache.flink.table.data.binary.BinaryStringDataFuryCodec_0
at java.lang.invoke.MemberName.checkForTypeAlias(MemberName.java:793) ~[?:1.8.0_372]
at java.lang.invoke.MemberName$Factory.resolve(MemberName.java:976) ~[?:1.8.0_372]
at java.lang.invoke.MemberName$Factory.resolveOrFail(MemberName.java:1000) ~[?:1.8.0_372]
at java.lang.invoke.MethodHandles$Lookup.resolveOrFail(MethodHandles.java:1386) ~[?:1.8.0_372]
at java.lang.invoke.MethodHandles$Lookup.findConstructor(MethodHandles.java:924) ~[?:1.8.0_372]
at org.apache.fury.serializer.Serializers.createSerializer(Serializers.java:127) ~[flash-serializer-v1.jar:?]
at org.apache.fury.serializer.Serializers.newSerializer(Serializers.java:104) ~[flash-serializer-v1.jar:?]
at org.apache.fury.resolver.ClassResolver$1.onSuccess(ClassResolver.java:941) ~[flash-serializer-v1.jar:?]
at org.apache.fury.resolver.ClassResolver$1.onSuccess(ClassResolver.java:938) ~[flash-se rializer-v1.jar:?]
at org.apache.fury.builder.JITContext.lambda$registerSerializerJITCallback$0(JITContext.java:94) ~[flash-serializer-v1.jar:?]
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_372]
... 2 more
```

### Anything Else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

--- a/java/fury-core/src/main/java/org/apache/fury/builder/CodecUtils.java
+++ b/java/fury-core/src/main/java/org/apache/fury/builder/CodecUtils.java
@@ -28,6 +28,7 @@ import org.apache.fury.reflect.TypeRef;
 import org.apache.fury.resolver.ClassResolver;
 import org.apache.fury.resolver.FieldResolver;
 import org.apache.fury.serializer.Serializer;
+import org.apache.fury.util.ClassLoaderUtils;
 import org.apache.fury.util.Preconditions;
 
 /** Codec util to create and load jit serializer class. */
@@ -100,7 +101,7 @@ public class CodecUtils {
       if (codeGenerator == null) {
         codeGenerator =
             CodeGenerator.getSharedCodeGenerator(
-                beanClassClassLoader, fury.getClass().getClassLoader());
+                ClassLoaderUtils.FuryJarClassLoader.getInstance(), beanClassClassLoader);
         // Hold strong reference of {@link CodeGenerator}, so the referent of `DelayedRef`
         // won't be null.
         classResolver.setCodeGenerator(loaders, codeGenerator);

--- a/java/fury-core/src/main/java/org/apache/fury/builder/CodecUtils.java
+++ b/java/fury-core/src/main/java/org/apache/fury/builder/CodecUtils.java
@@ -84,9 +84,26 @@ public class CodecUtils {
       beanClassClassLoader = fury.getClass().getClassLoader();
     }
     ClassResolver classResolver = fury.getClassResolver();
+    codeGenerator = getCodeGenerator(fury, beanClassClassLoader, classResolver);
+    ClassLoader classLoader =
+        codeGenerator.compile(
+            Collections.singletonList(compileUnit), compileState -> compileState.lock.lock());
+    String className = codecBuilder.codecQualifiedClassName(beanClass);
+    try {
+      return (Class<? extends Serializer<T>>) classLoader.loadClass(className);
+    } catch (ClassNotFoundException e) {
+      throw new IllegalStateException("Impossible because we just compiled class", e);
+    }
+  }
+
+  private static CodeGenerator getCodeGenerator(
+      Fury fury, ClassLoader beanClassClassLoader, ClassResolver classResolver) {
+    CodeGenerator codeGenerator;
     try {
       // generated code imported fury classes.
-      beanClassClassLoader.loadClass(Fury.class.getName());
+      if (beanClassClassLoader.loadClass(Fury.class.getName()) != Fury.class) {
+        throw new ClassNotFoundException();
+      }
       codeGenerator = classResolver.getCodeGenerator(beanClassClassLoader);
       if (codeGenerator == null) {
         codeGenerator = CodeGenerator.getSharedCodeGenerator(beanClassClassLoader);
@@ -107,14 +124,6 @@ public class CodecUtils {
         classResolver.setCodeGenerator(loaders, codeGenerator);
       }
     }
-    ClassLoader classLoader =
-        codeGenerator.compile(
-            Collections.singletonList(compileUnit), compileState -> compileState.lock.lock());
-    String className = codecBuilder.codecQualifiedClassName(beanClass);
-    try {
-      return (Class<? extends Serializer<T>>) classLoader.loadClass(className);
-    } catch (ClassNotFoundException e) {
-      throw new IllegalStateException("Impossible because we just compiled class", e);
-    }
+    return codeGenerator;
   }
 }

--- a/java/fury-core/src/main/java/org/apache/fury/util/ClassLoaderUtils.java
+++ b/java/fury-core/src/main/java/org/apache/fury/util/ClassLoaderUtils.java
@@ -33,6 +33,7 @@ import java.util.Map;
 import java.util.Objects;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.stream.Collectors;
+import org.apache.fury.Fury;
 import org.apache.fury.logging.Logger;
 import org.apache.fury.logging.LoggerFactory;
 import org.apache.fury.util.unsafe.DefineClass;
@@ -165,6 +166,33 @@ public class ClassLoaderUtils {
         }
       }
       return super.getResource(name);
+    }
+  }
+
+  /** A classloader to load Fury jar classes only. */
+  public static class FuryJarClassLoader extends URLClassLoader {
+    static {
+      ClassLoader.registerAsParallelCapable();
+    }
+
+    private static final ParentClassLoader LOADER =
+        new ParentClassLoader(Fury.class.getClassLoader());
+    private static final FuryJarClassLoader FURY_JAR_LOADER = new FuryJarClassLoader();
+
+    private FuryJarClassLoader() {
+      super(new URL[0]);
+    }
+
+    @Override
+    public Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+      if (name.startsWith(Fury.class.getPackage().getName())) {
+        return LOADER.loadClass(name, resolve);
+      }
+      return null;
+    }
+
+    public static FuryJarClassLoader getInstance() {
+      return FURY_JAR_LOADER;
     }
   }
 

--- a/java/fury-core/src/main/java/org/apache/fury/util/ClassLoaderUtils.java
+++ b/java/fury-core/src/main/java/org/apache/fury/util/ClassLoaderUtils.java
@@ -185,7 +185,7 @@ public class ClassLoaderUtils {
 
     @Override
     public Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
-      if (name.startsWith(Fury.class.getPackage().getName())) {
+      if (name.startsWith(Fury.class.getPackage().getName()) && !name.contains("test")) {
         return LOADER.loadClass(name, resolve);
       }
       return null;

--- a/java/fury-core/src/test/java/org/apache/fury/codegen/CodeGeneratorTest.java
+++ b/java/fury-core/src/test/java/org/apache/fury/codegen/CodeGeneratorTest.java
@@ -38,6 +38,7 @@ import org.apache.fury.collection.MultiKeyWeakMap;
 import org.apache.fury.test.bean.Foo;
 import org.apache.fury.util.ClassLoaderUtils;
 import org.apache.fury.util.ClassLoaderUtils.ByteArrayClassLoader;
+import org.apache.fury.util.ClassLoaderUtils.FuryJarClassLoader;
 import org.apache.fury.util.DelayedRef;
 import org.testng.Assert;
 import org.testng.annotations.Test;
@@ -70,6 +71,13 @@ public class CodeGeneratorTest {
     assertNotNull(
         sharedCodeGenerator2
             .get(new Object[] {getClass().getClassLoader(), Fury.class.getClassLoader()})
+            .get());
+    CodeGenerator.getSharedCodeGenerator(
+        FuryJarClassLoader.getInstance(), getClass().getClassLoader());
+    System.gc();
+    assertNotNull(
+        sharedCodeGenerator2
+            .get(new Object[] {FuryJarClassLoader.getInstance(), getClass().getClassLoader()})
             .get());
   }
 

